skimage/restoration/_nl_means_denoising.pyx

```diff
@@ -0,0 +1,596 @@
+import numpy as np
+from skimage import util
+cimport numpy as np
+cimport cython
+from libc.math cimport exp
+
+ctypedef np.float32_t DTYPE_t
+
+cdef eps = 1.e-8
+
+
+@cython.boundscheck(False)
+cdef inline float patch_distance_2d(DTYPE_t [:, :] p1,
+                                    DTYPE_t [:, :] p2,
+                                    DTYPE_t [:, ::] w, int s):
+    cdef int i, j
+    cdef int center = s / 2
+    # Check if central pixel is too different in the 2 patches
+    cdef float tmp_diff = p1[center, center] - p2[center, center]
+    cdef float init = w[center, center] * tmp_diff * tmp_diff
+    if init > 1:
+        return eps
+    cdef float distance = 0
+    for i in range(s):
+        # exp of large negative numbers will be 0, so we'd better stop
+        if distance > 5:
+            return eps
+        for j in range(s):
+            tmp_diff = p1[i, j] - p2[i, j]
+            distance += w[i, j] * tmp_diff * tmp_diff
+    distance = exp(- distance)
+    return distance
+
+
+@cython.boundscheck(False)
+cdef inline float patch_distance_2drgb(DTYPE_t [:, :, :] p1,
+                                       DTYPE_t [:, :, :] p2,
+                                       DTYPE_t [:, ::] w, int s):
+    cdef int i, j
+    cdef int center = s / 2
+    cdef int color
+    cdef float tmp_diff = 0
+    cdef float distance = 0
+    for i in range(s):
+        # exp of large negative numbers will be 0, so we'd better stop
+        if distance > 5:
+            return eps
+        for j in range(s):
+            for color in range(3):
+                tmp_diff = p1[i, j, color] - p2[i, j, color]
+                distance += w[i, j] * tmp_diff * tmp_diff
+    distance = exp(- distance)
+    return distance
+
+
+@cython.boundscheck(False)
+cdef inline float patch_distance_3d(DTYPE_t [:, :, :] p1,
+                                    DTYPE_t [:, :, :] p2,
+                                    DTYPE_t [:, :, ::] w, int s):
+    cdef int i, j, k
+    cdef float distance = 0
+    cdef float tmp_diff
+    for i in range(s):
+        # exp of large negative numbers will be 0, so we'd better stop
+        if distance > 5:
+            return eps
+        for j in range(s):
+            for k in range(s):
+                tmp_diff = p1[i, j, k] - p2[i, j, k]
+                distance += w[i, j, k] * tmp_diff * tmp_diff
+    distance = exp(- distance)
```

Comment:
unary operators should not be separated by a space from their argument: `-distance`